### PR TITLE
Add broadcast of block to peers

### LINKED ISSUE #3: Block propagation
Im making small changes to both Block and MerkleTree classes and mostly changing started, check_transactions, on_block and finalize_and_broadcast_block functions

### PR DESCRIPTION
**Pull request by totisama**
With this changes we are able to broadcast the block to other peers.
I added the merkle_hash variable to block so we can have the hash without needing to serialize the MerkleTree instance.

### CODE REVIEW

File: block.py [hunk 1]
```diff
@@ -1,33 +1,34 @@
...
         self.merkle_tree = MerkleTree()
+        self.merkle_hash = merkle_hash if merkle_hash != '' else ''
 
     def add_transaction(self, transaction):
-        self.transactions.append(transaction)
-        serialized_data = self.serialize(transaction)
-        transaction_hash = hashlib.sha256(serialized_data).hexdigest()
+        # We serialize the transaction before appending it to the array
+        # because there seems to be an ipv8 problem when it tries to serialize the whole array
+        serialized_transaction = default_serializer.pack_serializable(transaction)
```
Comment (KarinaGanieva-sl) [inline block.py:21]: here we need to exclude some field like ttl from transaction before serializing, but I will fix it with my pr

File: block.py [hunk 2]
```diff
@@ -1,33 +1,34 @@
...
-        })
+    def update_tree(self):
+        self.merkle_tree.recalculate_tree()
+        self.merkle_hash = self.merkle_tree.get_root_hash()
 
-    def __reduce__(self):
-        # Customizing pickling for Block class
-        return (self.__class__, (self.previous_hash, self.merkle_tree))
\ No newline at end of file
+    def get_merkle_hash(self) -> bytes:
```
Comment (KarinaGanieva-sl) [inline block.py]: merkle_hash is a string and here you returns bytes, do we need to have a string for hashes or bytes?

File: merkle_tree.py [hunk 3]
```diff
@@ -12,12 +12,13 @@ def __init__(self, hash):
...
         self.leaves = []
-        self.root = None
+        self.root: MerkleNode = None
 
     def add_leaf(self, leaf_hash):
         new_leaf = MerkleNode(leaf_hash)
         self.leaves.append(new_leaf)
-        self.recalculate_tree()
+        # Do we need to calculate the tree everytime we add a leaf?
+        # self.recalculate_tree()
```
Comment (KarinaGanieva-sl) [inline merkle_tree.py]: no, I think when we have 10 txs - only then we can build a tree

File: main.py [hunk 4]
```diff
@@ -49,15 +51,19 @@ def __init__(self, settings: CommunitySettings) -> None:
         self.current_block = Block('0')  # Current working block
 
         self.add_message_handler(Transaction, self.on_transaction)
-        self.add_message_handler(BlockMessage, self.on_block)
+        self.add_message_handler(BlockMessage, self.receive_block)
 
     def started(self, id) -> None:
+        print('Community started', self.get_peer_id(self.my_peer))
```
Comment (KarinaGanieva-sl) [inline main.py]: why don't you use logging here

File: main.py [hunk 5]
```diff
@@ -134,40 +142,45 @@ async def on_transaction(self, peer: Peer, payload: Transaction) -> None:
...
+        new_block_hash = self.current_block.get_merkle_hash()
         print(f'New block hash: {new_block_hash}')
         logging.info(f'New block hash: {new_block_hash}')
         self.blocks.append(self.current_block)
-        serialized_block = pickle.dumps(self.current_block)
-        self.current_block = Block(new_block_hash)
 
-        self.broadcast_new_block(new_block_hash, serialized_block)
+        self.broadcast_block(new_block_hash, self.current_block)
+        self.current_block = Block(new_block_hash)
```
Comment (KarinaGanieva-sl) [inline main.py:171]: here also "Expected type 'str', got 'bytes' instead"

File: main.py [hunk 6]
```diff
@@ -93,23 +99,25 @@ def create_transaction(self):
...
+        # WIP: get every known peer, not only the ones im connected to
+        peers = self.get_peers()
+        peers.append(self.my_peer)
+        selectedPeer = random.choice(peers)
+
+        if not selectedPeer.mid == self.my_peer.mid:
+            return
 
         for tx in self.pending_txs:
-            if self.balances[tx.sender] - tx.amount >= 0:
```
Comment (KarinaGanieva-sl) [inline main.py:102]: Do you want to add this part to on_transaction?

File: main.py [hunk 7]
```diff
@@ -93,23 +99,25 @@ def create_transaction(self):
...
 
-    def check_transactions(self):
-        print(f'[Node {self.get_peer_id(self.my_peer)}] Checking transactions')
-        logging.info(f'[Node {self.get_peer_id(self.my_peer)}] Checking transactions')
 
+    def block_creation(self):
+        # WIP: get every known peer, not only the ones im connected to
+        peers = self.get_peers()
+        peers.append(self.my_peer)
+        selectedPeer = random.choice(peers)
```
Comment (KarinaGanieva-sl) [inline main.py:106]: not sure about this part, can we discuss a code with Bulat tomorrow?

File: block.py (hunk 2 above)
Comment (totisama) [reply block.py]: thanks, now its fixed

File: merkle_tree.py (hunk 3 above)
Comment (totisama) [reply merkle_tree.py]: perfect, I will remove it then

File: main.py (hunk 4 above)
Comment (totisama) [reply main.py]: I forgot to remove that, thanks

File: main.py (hunk 5 above)
Comment (totisama) [reply main.py:171]: thanks, now it should be good

File: main.py (hunk 6 above)
Comment (totisama) [reply main.py:102]: Not to on transaction but to when we receive a block
In todays class Bulat explained that when we receive a block we have to do a stateful and stateless validation of the block, so this validation of the tx's would be at this point

File: main.py (hunk 7 above)
Comment (totisama) [reply main.py:106]: yeah, sure

**Pull request merged**